utils/merge.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// Merge will return a new struct/map/slice of the same type as base and patch, with patch merged into base.
+// Specifically, patch's values will be preferred except when patch's value is `nil`.
+// Restrictions/guarantees:
+//   - base and patch will not be modified
+//   - base and patch can be pointers or values
+//   - base and patch must be the same type
+//   - if slices are different, they will be merged according to the following rules:
```

Comment:
I really like the flexibility to support these use cases -- that being said (and speaking selfishly ;)), merging of `*model.Config` should replace slices and maps atomically vs. merging their contents. If we want to presere these semantics, could we expose a configuration to control this?
```
type SliceMergeConfig int

const (
    ReplaceSlices                SliceMergeConfig = iota
    MergeSlicesOverwrite
)

type MapMergeConfig int

const (
    ReplaceMaps                MapMergeConfig = iota
    MergeMapsOverwrite
)

type MergeConfig struct {
    SliceMergeConfig SliceMergeConfig
    MapMergeConfig MapMergeConfig
}

func Merge(base interface{}, patch interface{}, config *MergeConfig) (interface{}, error) {
    // ...
}
```




This would also fit well into a planned extension for this library: callback functions to handle specific cases. For example, I'm presently annotating `*model.Config` with [StructTags](https://golang.org/pkg/reflect/#StructTag) to configure which fields can be modified by a restricted system admin. I could extend this config parameter with a `MergeCheckCallback func(base, patch reflect.Value, tag reflect.StructTag) bool` to effectively reuse your recursion for a customized patch case.

Reply:
That sounds great, I added it verbatim -- I'm still getting a handle on the codebase, so I'm not sure if this function-specific config belongs in the merge.go file? (That's where I put it for now.) Please let me know.